### PR TITLE
Remove unused diagnostic 2346

### PR DESCRIPTION
**Pull request by Andarist**
I noticed that this code is never executed

### CODE REVIEW

**Comment by fatcerberus** [comment]
This is interesting, because googling TS2346 actually turns up results (including at least one issue here) but apparently the error message used to be `Supplied parameters do not match any signature of call target.`

**Comment by fatcerberus** [comment]
Also possibly relevant: https://github.com/microsoft/TypeScript/issues/17032

**Comment by Andarist** [comment]
I think that the equivalent diagnostics are just reported by different code paths nowadays. So those errors didn't actually go away.

**Comment by fatcerberus** [comment]
Yeah, probably fixing #17032 is what made this specific diagnostic unused

**Comment by RyanCavanaugh** [comment]
ELI5, how do you know this code "never" runs? It appears reachable to me.

**Comment by Andarist** [comment]
1. I noticed that this diagnostic doesn't appear anywhere in the test suite
2. after removing this branch of code the whole test suite still passed. 
3. I replaced the removed code with `if (!candidates.length) { throw new Error('here!') }`, reran tests and they still passed

**Comment by Andarist** [comment]
I've done an additional call hierarchy analysis for this

1. `candidates.length` is always 0 at the start
2. `reorderCandidates` fills in `candidates` based on `signatures` and after that `candidates.length === signatures.length` should always be true
3. so the only way to end up with this error is to have `signatures.length === 0`
4. all but one path leading to this have dedicated branches for their respective `signatures.length` (like `if (constructSignatures.length)` in `resolveNewExpression` etc) and if the list is empty then a dedicated error is added to diagnostic and `resolveCall` is not called at all
5. the only "unguarded" path is dedicated to `super()` but I believe that `resolveCall` can't be called in there either with `baseConstructors.length === 0`. I both tried different inputs and couldn't make it work and also did an additional analysis here
- this depends on `checkSuperExpression`'s result - if it returns any/error type then we skip the `resolveCall` (so we can't reach the code that is being removed by this PR)
- that, in turn, depends on `baseClassType` obtained through `getBaseTypes`/`.resolvedBaseTypes`
- those `.resolvedBaseTypes` can actually be empty but that just means that in such a case we return `errorType` from `checkSuperExpression` (since `!baseClassType`)
- so, in fact, we need a `baseClassType` to move forward to `resolveCall` and I think **that** this implies having at least one valid constructor which also implies that we won't pass an empty `signatures` list to `resolveCall`

> Incredibly flattered that you think our test suites are that exhaustive, but it's really not the case 😅.

A guy saying 'I have total faith in you'

**Comment by DanielRosenwasser** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test tsserver top100
@typescript-bot run dt

**Comment by RyanCavanaugh** [comment]
I'd be OK replacing this with `Debug.assert(..., "Revert #54442 and add a testcase with whatever triggered this");` and then doing a full remove in 5.3 😅

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test tsserver top100
@typescript-bot run dt

**Comment by rbuckton** [comment]
I found something that triggers the debug assertion: https://github.com/microsoft/TypeScript/issues/55499

**Comment by Andarist** [comment]
"Post mortem" of what went wrong with the initial investigation is that this didn't hold true:
> so, in fact, we need a baseClassType to move forward to resolveCall and I think that this implies having at least one valid constructor which also implies that we won't pass an empty signatures list to resolveCall

I added further notes on how there might be no valid constructor even in the presence of a `baseClassType` here:
https://github.com/microsoft/TypeScript/pull/55506#discussion_r1304947744

**Comment by RyanCavanaugh** [comment]
Two years later, we received another way to hit this (via https://github.com/microsoft/TypeScript/issues/61438#issuecomment-2902575687)
```ts
export class SomeBaseClass { }
export interface SomeInterface { }
export interface MergedClass extends SomeInterface { }
export class MergedClass extends SomeBaseClass<any> {
	public constructor() {
		super();
	}
}
```

**Comment by RyanCavanaugh** [comment]
Really a cautionary tale here; I shouldn't have approved this. It's not just the time to author and review PRs, it's the time that [users spend](https://github.com/microsoft/TypeScript/issues/61438#issuecomment-2902257547) tracking these things down -- those people didn't sign up for this work and no one got any tangible benefit from this not-actually-dead code being removed in the interim.
